### PR TITLE
Add generic parameters to functions

### PR DESCRIPTION
**Pull request by Override-6**
This pull requests adds the support of generic functions to the analyzer and compiler.
- Compiler and VM can now handle qword and byte box/unboxing
- `Typing` can now contain an optional name for registered types

### CODE REVIEW

**Review by syldium — CHANGES_REQUESTED** [review]
No issue is detected in
```
fun one[T]() -> T = 1
echo one[String]()
```

File: analyzer/src/steps/typing.rs [hunk 1]
```diff
@@ -150,31 +162,55 @@ fn apply_types_to_source(
...
+            let base_chunk = forward_declaration
+                .unwrap_or_else(|| type_function_signature(func, exploration, links, diagnostics));
+
+            let expected_return_type =
+                if let Some(return_type_annotation) = func.return_type.as_ref() {
+                    // An explicit return type is present, check it against all the return types.
+                    resolve_type_annotation(exploration, links, return_type_annotation, diagnostics)
+                } else {
+                    UNIT
+                };
```
Comment (syldium) [inline analyzer/src/steps/typing.rs:177]:
```suggestion
            let expected_return_type = base_chunk.return_type;
```

File: analyzer/src/steps/typing.rs [hunk 2]
```diff
@@ -187,6 +223,68 @@ fn apply_types_to_source(
     }
 }
 
+/// create a basic chunk from a function declaration
+/// type its parameters, type parameters and return type
+fn type_function_signature(
```
Comment (syldium) [inline analyzer/src/steps/typing.rs:228]: This function belongs to `typing/function.rs`.

File: analyzer/src/steps/typing.rs [hunk 3]
```diff
@@ -150,31 +162,55 @@ fn apply_types_to_source(
     exploration.prepare();
     match expr {
         Expr::FunctionDeclaration(func) => {
-            let mut params = Vec::new();
-            for param in &func.parameters {
-                let param = type_parameter(exploration, param, links, diagnostics);
-                exploration.ctx.push_local_typed(source_id, param.ty);
-                params.push(param);
-            }
+            // is some if this chunk was forward declared
```
Comment (syldium) [inline analyzer/src/steps/typing.rs:165]:
```suggestion
            // Take any previous forward declaration if present.
```

File: analyzer/src/steps/typing.rs [hunk 4]
```diff
@@ -260,29 +358,33 @@ fn ascribe_assign(
...
+                        exploration,
+                        links,
+                        diagnostics,
+                        &binary,
+                        state.with_local_type(),
+                    )
+                }
+            }
+        };
+    }
```
Comment (syldium) [inline analyzer/src/steps/typing.rs:387]:
```suggestion
    let actual_type_ref = links
        .env()
        .get_raw_symbol(assign.left.segment())
        .map(|symbol| {
            exploration
                .ctx
                .get(links.relations, links.source, symbol)
                .unwrap()
                .type_ref
        });
    let rhs = match assign.operator {
        AssignOperator::Assign => ascribe_types(
            exploration,
            links,
            diagnostics,
            &assign.value,
            state.with_local_type().with_expected_type(actual_type_ref),
        ),
        operator => {
            let binary = Expr::Binary(BinaryOperation {
                left: assign.left.clone(),
                op: BinaryOperator::try_from(operator).expect("Invalid assign operator"),
                right: assign.value.clone(),
            });
            ascribe_types(
                exploration,
                links,
                diagnostics,
                &binary,
                state.with_local_type(),
            )
        }
    };
```
Avoid big macros that lose syntax highlighting and that can be avoided here.

File: analyzer/src/steps/typing/coercion.rs [hunk 5]
```diff
@@ -36,28 +37,64 @@ pub(super) fn convert_description(
...
-    if *rhs == Type::Nothing {
+    // apply the `A U Nothing => A` rule only if `A` is a base type
+    if is_base_type && *rhs == Type::Nothing {
         return Ok(assign_to);
     }
 
+    if let (Type::Instantiated(base_left, params_lhs), Type::Instantiated(base_right, params_rhs)) =
+        (lhs, rhs)
+    {
+        if base_left == base_right {
```
Comment (syldium) [inline analyzer/src/steps/typing/coercion.rs:64]:
```suggestion
        if is_compatible(exploration, *base_left, *base_right)
```

File: analyzer/src/steps/typing/exploration.rs [hunk 6]
```diff
@@ -143,17 +154,17 @@ impl Exploration<'_> {
...
             _ => type_id,
         })
     }
 
-    pub(super) fn get_external_env<'a>(
-        &'a self,
-        from_env: &'a Environment,
+    pub(super) fn get_external_env<'b>(
+        &'b self,
+        from_env: &'b Environment,
```
Comment (syldium) [inline analyzer/src/steps/typing/exploration.rs:165]:
```suggestion
    pub(super) fn get_external_env(
        &'a self,
        from_env: &'a Environment,
```

File: analyzer/src/steps/typing/exploration.rs [hunk 7]
```diff
@@ -143,17 +154,17 @@ impl Exploration<'_> {
...
         to_symbol: ResolvedSymbol,
-    ) -> Option<&'a Environment> {
+    ) -> Option<&'b Environment> {
```
Comment (syldium) [inline analyzer/src/steps/typing/exploration.rs:167]:
```suggestion
    ) -> Option<&'a Environment> {
```

File: analyzer/src/types/engine.rs [hunk 8]
```diff
@@ -127,6 +141,11 @@ impl TypedEngine {
         self.entries.get(id.0)?.as_ref()
     }
 
+    pub fn take_user(&mut self, id: SourceId) -> Option<Chunk> {
+        self.entries.push(None);
+        self.entries.swap_remove(id.0)
```
Comment (syldium) [inline analyzer/src/types/engine.rs:146]:
```suggestion
        self.entries.get_mut(id.0)?.take()
```

File: compiler/src/bytecode.rs [hunk 9]
```diff
@@ -118,6 +119,14 @@ impl<'a> Instructions<'a> {
         self.emit_code(pop_opcode);
     }
 
+    pub fn emit_box(&mut self, ty: TypeRef) {
+        match ty {
+            types::EXITCODE | types::BOOL => self.emit_code(Opcode::BoxByte),
+            types::INT | types::FLOAT => self.emit_code(Opcode::BoxQWord),
+            _ => panic!("Can only box int, float, exitcode and boolean primitives"),
```
Comment (syldium) [inline compiler/src/bytecode.rs:126]:
```suggestion
            _ => { /* Objects are already boxed */ }
```

File: compiler/src/emit/invoke.rs [hunk 10]
```diff
@@ -139,8 +139,23 @@ pub fn emit_function_invocation(
...
+
+    let function_definition = ctx.get_function(function_call.reef, func_source).unwrap();
+
+    for (arg, parameter) in function_call
+        .arguments
+        .iter()
+        .zip(&function_definition.parameters)
+    {
         emit(arg, instructions, ctx, cp, locals, state);
+        // the parameter is an object but argument is not : may be an argument passed to a generic parameter
```
Comment (syldium) [inline compiler/src/emit/invoke.rs:154]:
```suggestion
        // The parameter is an object but the argument isn't: may be an argument passed to a generic parameter
```

File: compiler/src/emit/invoke.rs [hunk 11]
```diff
@@ -188,6 +204,9 @@ pub fn emit_function_invocation(
         // in order to maintain the stack's size, we instantly pop
         // the stack if the value isn't used later in the code
         instructions.emit_pop(return_type_size);
+    } else if return_type.is_obj() && !call_return_type.is_obj() {
+        // the function's declared return type is an object but call return type is not : its a boxed return value
```
Comment (syldium) [inline compiler/src/emit/invoke.rs:208]:
```suggestion
        // The function's declared return type is an object but the call return type is not: its a boxed return value
```

**Review by syldium — CHANGES_REQUESTED** [review]

File: analyzer/src/steps/typing.rs (hunk 4 above)
Comment (Override-6) [reply analyzer/src/steps/typing.rs:387]: it causes code duplication

Comment (syldium) [reply analyzer/src/steps/typing.rs:387]: Macros are just expanded. The point here is that there is no need for two calls to a `ascribe_rhs` macro.

File: analyzer/src/steps/typing.rs [hunk 12]
```diff
@@ -64,14 +67,25 @@ pub fn apply_types(
...
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Default)]
+enum ExpressionValue {
+    /// The value of the expression is not used
+    #[default]
+    Unused,
+    /// The value of the expression is used but its type is not specified
+    Unspecified,
+    /// Thevalue oftheexpression is used and is of expected type
```
Comment (syldium) [inline analyzer/src/steps/typing.rs:77]:
```suggestion
    /// A type is expected and used.
```

**Review by syldium — APPROVED** [review]

**Pull request merged**